### PR TITLE
DB collections monitoring

### PR DESCRIPTION
**Pull request by alesstimec**
Added a collector that reports the size and object count for all collections in a database.

### CODE REVIEW

File: collection_size.go [hunk 1]
```diff
@@ -22,6 +22,94 @@ type stats struct {
 	Size  int32 `bson:"size"`
 }
 
+func NewDBCollectionsSizeCollector(namespace string, db *mgo.Database) *DBCollectionsSizeCollector {
```
Comment (cmars) [inline collection_size.go]: godoc please

File: collection_size.go [hunk 2]
```diff
@@ -22,6 +22,94 @@ type stats struct {
...
+		countDesc: prometheus.NewDesc(
+			prometheus.BuildFQName(namespace, "database", fmt.Sprintf("%s_collection_count", db.Name)),
+			"collection object count",
+			[]string{"collection"},
+			nil),
+		db: dbWithCopiedSession,
+	}
+}
+
+type DBCollectionsSizeCollector struct {
```
Comment (cmars) [inline collection_size.go]: godoc please

File: collection_size.go [hunk 3]
```diff
@@ -22,6 +22,94 @@ type stats struct {
...
+	countDesc *prometheus.Desc
+
+	mu     sync.Mutex
+	db     *mgo.Database
+	closed bool
+}
+
+var _ prometheus.Collector = (*DBCollectionsSizeCollector)(nil)
+
+func (u *DBCollectionsSizeCollector) Close() {
```
Comment (cmars) [inline collection_size.go]: godoc?

**Review by cmars — APPROVED** [review]
Do we need `CollectionSizeCollector` anymore, now that we'll have a `DBCollectionSizeCollector`? If we want both, can we reduce code duplication without much complexity? If not, let's remove `CollectionSizeCollector`.

Otherwise, fix it & ship it!

File: collection_size.go [hunk 4]
```diff
@@ -22,6 +22,97 @@ type stats struct {
 	Size  int32 `bson:"size"`
 }
 
+// NewDBCollectionsSizeCollector returns a new collector that monitors sizes and object counts
+// of all collections in the specified database.
+func NewDBCollectionsSizeCollector(namespace string, db *mgo.Database) *dbCollectionsSizeCollector {
```
Comment (mattyw) [inline collection_size.go]: why not just a db string?

File: collection_size.go [hunk 5]
```diff
@@ -22,6 +22,97 @@ type stats struct {
...
+	dbWithCopiedSession := s.DB(db.Name)
+	return &dbCollectionsSizeCollector{
+		sizeDesc: prometheus.NewDesc(
+			prometheus.BuildFQName(namespace, "database", fmt.Sprintf("%s_collection_size_bytes", db.Name)),
+			"collection size in bytes",
+			[]string{"collection"},
+			nil,
+		),
+		countDesc: prometheus.NewDesc(
+			prometheus.BuildFQName(namespace, "database", fmt.Sprintf("%s_collection_count", db.Name)),
```
Comment (mattyw) [inline collection_size.go]: collection_doc_count

File: collection_size.go [hunk 6]
```diff
@@ -22,6 +22,97 @@ type stats struct {
...
+			"collection object count",
```
Comment (mattyw) [inline collection_size.go]: doc count

File: collection_size.go [hunk 7]
```diff
@@ -22,6 +22,97 @@ type stats struct {
...
+func (u *dbCollectionsSizeCollector) Describe(c chan<- *prometheus.Desc) {
+	c <- u.sizeDesc
+	c <- u.countDesc
+}
+
+// Collect implements the prometheus.Collector interface.
+func (u *dbCollectionsSizeCollector) Collect(ch chan<- prometheus.Metric) {
+	var session *mgo.Session
+	u.mu.Lock()
+	if !u.closed {
```
Comment (mattyw) [inline collection_size.go]:
```
if u.close {
    return
}
```

File: collection_size.go (hunk 4 above)
Comment (alesstimec) [reply collection_size.go]: then we'd need a string and a session..

Comment (mattyw) [reply collection_size.go]: I'm inclined to think that's better. We don't usually pass around mgo.Database so I'd imagine needing to create one just to pass to this function. But we don't actually use it, the first thing we do is create a new db using a copied session and the name from this one

Comment (alesstimec) [reply collection_size.go]: i think passing in the db is ok.. unless you have a very strong reason not to do so..

Comment (mattyw) [reply collection_size.go]: I don't care enough to block landing it, But you need two things to make a database, session and a name string. But we never actually use it. That means clients have to make one - but it's never used: https://github.com/cloud-green/monitoring/pull/9/files#diff-be82fd1dce085c2ac204051ff78a9000R27

I'll let your conscience decide

Comment (alesstimec) [reply collection_size.go]: @mattyw  my conscience is clear :)

**Review by mattyw — APPROVED** [review]

**Comment by alesstimec** [comment]
$$merge$$

**Pull request merged**